src/script/microbit-interfacing/MicrobitSerial.ts

```diff
@@ -133,25 +141,43 @@ export class MicrobitSerial implements MicrobitConnection {
           buttons: true,
         });
         await this.usb.serialWrite(startCmd.message);
+        const periodicMessagePromise = new Promise<void>((resolve, reject) => {
+          onPeriodicMessageRecieved = resolve;
+          setTimeout(() => {
+            onPeriodicMessageRecieved = undefined;
+            reject(new Error('Failed to receive data from remote micro:bit'));
+          }, 500);
+        });
         await this.sendCmdWaitResponse(startCmd);
+        if (this.isReconnect) {
+          await periodicMessagePromise;
+        } else {
+          periodicMessagePromise.catch(async e => {
+            logError('Failed to initialise serial protocol', e);
+            await this.disconnectInternal(false, 'remote');
+            this.isConnecting = false;
+          });
+        }
       }
 
       stateOnAssigned(DeviceRequestStates.INPUT, this.usb.getModelNumber());
       stateOnReady(DeviceRequestStates.INPUT);
       logMessage('Serial successfully connected');
     } catch (e) {
       logError('Failed to initialise serial protocol', e);
-      stateOnFailedToConnect(DeviceRequestStates.INPUT);
-      await this.usb.stopSerial();
+      let reconnectHelp: ConnectionType = 'remote';
+      if (typeof e === 'string' && e.includes('Handshake')) {
```

Comment:
This has ended up kinda gross.

Reply:
Maybe we throw a specific exception type? Either use instanceof to identify or perhaps have a ConnectionError exception with error codes? I guess you'll need instanceof either way.

Reply:
Yeah, I'll create a PR for this as it's currently a bit horrid and fragile.